File: repos/kotlin-inventory/src/main/kotlin/com/optarena/inventory/service/InventoryService.kt
package com.optarena.inventory.service

import com.optarena.inventory.model.Movement
import com.optarena.inventory.repository.ItemRepository
import com.optarena.inventory.repository.MovementRepository
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.web.server.ResponseStatusException

@Service
class InventoryService(
    private val itemRepository: ItemRepository,
    private val movementRepository: MovementRepository
) {
    fun applyMovement(itemId: Long, type: String, quantity: Int, note: String?): Movement {
        val item = itemRepository.findById(itemId)
            ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "item not found")
        if (type != "IN" && type != "OUT") {
            throw ResponseStatusException(HttpStatus.BAD_REQUEST, "type must be IN or OUT")
        }
        if (quantity <= 0) {
            throw ResponseStatusException(HttpStatus.BAD_REQUEST, "quantity must be a positive integer")
        }
        val delta = if (type == "IN") quantity else -quantity
        val newQuantity = item.quantity + delta
        if (newQuantity < 0) {
            throw ResponseStatusException(HttpStatus.BAD_REQUEST, "movement would take quantity below zero")
        }
        item.quantity = newQuantity
        itemRepository.save(item)
        return movementRepository.save(Movement(itemId = itemId, type = type, quantity = quantity, note = note))
    }
}

File: repos/express-js-inventory/src/services/inventoryService.js
const itemStore = require("../store/itemStore");
const movementStore = require("../store/movementStore");

const VALID_TYPES = new Set(["IN", "OUT"]);

/**
 * Applies a stock movement to an item: IN adds to quantity, OUT subtracts.
 * Returns { error } on validation failure, or { movement, item } on success.
 * Rejects an OUT movement that would take quantity below zero.
 */
function applyMovement(item, { type, quantity, note }) {
  if (!VALID_TYPES.has(type)) {
    return { error: "type must be IN or OUT" };
  }
  if (typeof quantity !== "number" || !Number.isInteger(quantity) || quantity <= 0) {
    return { error: "quantity must be a positive integer" };
  }

  const delta = type === "IN" ? quantity : -quantity;
  const newQuantity = item.quantity + delta;
  if (newQuantity < 0) {
    return { error: "movement would take quantity below zero" };
  }

  itemStore.updateItem(item.id, { quantity: newQuantity });
  const movement = movementStore.createMovement({
    itemId: item.id,
    type,
    quantity,
    note: typeof note === "string" ? note : null,
    at: new Date().toISOString(),
  });
  return { movement, item: itemStore.getItem(item.id) };
}

module.exports = { applyMovement };
